Source/CarthageKit/Xcode.swift

```diff
@@ -678,18 +719,151 @@ public func buildScheme( // swiftlint:disable:this function_body_length cyclomat
+										case let .standardError(data):
+											return SignalProducer(value: .standardError(data))
+										case let .success(m):
+											return SignalProducer(m).flatMap(.merge) { target, settings in
+												copyBuildProductIntoDirectory(settings.productDestinationPath(in: directoryURL.appendingPathComponent("UIKitForMac")), settings)
+													.zip(with: SignalProducer<String, CarthageError>(value: target))
+												}
+												.collect()
+												.flatMap(.merge) { macOSTargetsDestinationULRs in
+													return iPhoneSimulatorSettingsProducer.flatMapTaskEvents(.concat) { s in
+														return SignalProducer(s).flatMap(.merge) { (target, settings) -> SignalProducer<(URL, String), CarthageError>  in
+															copyBuildProductIntoDirectory(settings.productDestinationPath(in: directoryURL.appendingPathComponent("iOSSimulator")), settings)
+															.zip(with: SignalProducer<String, CarthageError>(value: target))
+															}
+															.collect()
+															.flatMap(.merge) { simualtorTargetsDestinationULRs -> SignalProducer<Triplet, CarthageError> in
+																let triplet: Triplet = ((iphoneTargetsDestinationULRs, d),(macOSTargetsDestinationULRs, m),(simualtorTargetsDestinationULRs, s))
+																return SignalProducer<Triplet, CarthageError>(value: triplet)
+															}
+														}
+												}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+				.flatMapTaskEvents(.concat) { dT, mT, sT in
+
+					let (dURLs, deviceSettings) = dT
+					let (mURLs, _) = mT
+					let (sURLs, _) = sT
+
+
+					let reducer: (inout [String : Set<URL>], (URL, String)) -> () = { acc, tuple in
+						if acc[tuple.1] == nil {
+
+							acc[tuple.1] = Set([tuple.0])
+						}
+						else {
+							acc[tuple.1]!.update(with: tuple.0)
+						}
+					}
+
+					let deviceTargetsAndProductsDictionary = dURLs.reduce(into: [String : Set<URL>](), reducer)
+					let macTargetsAndProductsDictionary = mURLs.reduce(into: [String : Set<URL>](), reducer)
+					let simulatorTargetsAndProductsDictionary = sURLs.reduce(into: [String : Set<URL>](), reducer)
+
+					var allPlatformsTargetURLs = [String : Set<URL>]()
+
+					for (target, urls) in deviceTargetsAndProductsDictionary {
+
+						allPlatformsTargetURLs[target] = urls
+							.union(macTargetsAndProductsDictionary[target]!)
+							.union(simulatorTargetsAndProductsDictionary[target]!)
+					}
+
+					return SignalProducer(allPlatformsTargetURLs)
+						.flatMap(.merge) { target, urls in
+							let outputURL = deviceSettings[target]!
+							.xcFrameworkWrapperName
+							.map(folderURL.appendingPathComponent)
+
+							let adaptedURLs = Result<[URL], CarthageError>(Array(urls))
+
+							return SignalProducer(outputURL.fanout(adaptedURLs))
+								.flatMap(.merge) { createXCFramework($0.1, $0.0) }
+								.then(SignalProducer(result: outputURL))
+					}
+				}
```

Comment:
This code is hard to read in the diff on GitHub. It's long and has a lot of nesting. 🙈 Can we do something to clean it up?